packages/authentication/test/acceptance/basic-auth.ts

```diff
@@ -0,0 +1,206 @@
+    });
+  });
+
+  it('returns error for invalid credentials', () => {
+    return whenIMakeRequestTo(app).then(client => {
+      const credential =
+        users.list.Simpson.profile.id + ':' + 'invalid';
+      const hash = new Buffer(credential).toString('base64');
+      return client.get('/whoAmI')
+        .set('Authorization', 'Basic ' + hash)
+        .expect(401);
+    });
+  });
+
+  function givenUserRespository() {
+    users = new UserRepository({
+      joe : {profile: {id: 'joe'}, password: '12345'},
+      Simpson: {profile: {id: 'sim123'}, password: 'alpha'},
+      Flintstone: {profile: {id: 'Flint'}, password: 'beta'},
+      George: {profile: {id: 'Curious'}, password: 'gamma'},
+    });
+  }
+
+  function givenAnApplication() {
+    app = new Application();
+    app.bind('application.name').to('SequenceApp');
+  }
+
+  function givenControllerInApp() {
+    const apispec = givenOpenApiSpec()
+      .withOperation('get', '/whoAmI', {
+        'x-operation-name': 'whoAmI',
+        responses: {
+          '200': {
+            type: 'string',
+          },
+        },
+      })
+      .build();
+
+    @api(apispec)
+    class MyController {
+      constructor(@inject('authentication.user') private user: UserProfile) {}
+
+      @authenticate('BasicStrategy')
+      async whoAmI() : Promise<string> {
+        return this.user.id;
+      }
+    }
+    app.controller(MyController);
+  }
+
+  function givenAuthenticatedSequence() {
+    class MySequence {
+      constructor(
+        @inject('findRoute') protected findRoute: FindRoute,
+        @inject('getFromContext') protected getFromContext: getFromContext,
+        @inject('invokeMethod') protected invoke: InvokeMethod,
+        @inject('bindElement') protected bindElement: bindElement,
+      ) {}
+
+      async run(req: ParsedRequest, res: ServerResponse) {
+        try {
+          const {
+            controller,
+            methodName,
+            spec: routeSpec,
+            pathParams,
+          } = this.findRoute(req);
+
+          // Resolve authenticate() from AuthenticationProvider
+          const authenticate: AuthenticateFn =
+            await this.getFromContext(BindingKeys.Authentication.PROVIDER);
+
+          // Authenticate
+          const user: UserProfile = await authenticate(req);
+
+          // User is expected to be returned or an exception should be thrown
+          if (user) this.bindElement('authentication.user', user);
+          else throw new HttpErrors.InternalServerError('auth error');
+
+          // Authentication successful, proceed to invoke controller
+          const args = await parseOperationArgs(req, routeSpec, pathParams);
+          const result = await this.invoke(controller, methodName, args);
+          writeResultToResponse(res, result);
+        } catch (err) {
+          this.sendError(res, req, err);
+          return;
+        }
+      }
+      sendError(res: ServerResponse, req: ServerRequest, err: HttpError) {
```

Comment:
Hmm, `sendError` should be injectable so that Sequence classes don't have to (re)implement it again and again. That's out of scope of this pull request though.